// monkeysjs/dist/monkeysjs.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var monkeysjs_esm_exports = {};
__export(monkeysjs_esm_exports, {
  HttpError: () => ie,
  RequestState: () => K,
  WebSocketState: () => N,
  autoInit: () => ve,
  batch: () => Qe,
  camelCase: () => zt,
  cancellable: () => xt,
  capitalize: () => qt,
  component: () => lt,
  computed: () => ae,
  cookies: () => ce,
  createApp: () => _e,
  createClient: () => xe,
  createEventEmitter: () => It,
  createIndexedDB: () => bt,
  createMemoryStorage: () => Oe,
  createQueue: () => Pt,
  createWebSocketClient: () => it,
  debounce: () => Et,
  deepClone: () => V,
  deepMerge: () => Ee,
  default: () => Se,
  directive: () => v,
  effect: () => X,
  escapeHtml: () => _t,
  formatBytes: () => Bt,
  formatNumber: () => Ut,
  get: () => $t,
  has: () => Ct,
  http: () => ot,
  isObject: () => ue,
  isReactive: () => ge,
  isRef: () => ye,
  kebabCase: () => Ft,
  localStorage: () => ze,
  memoize: () => Nt,
  omit: () => Lt,
  parseQuery: () => jt,
  pick: () => Tt,
  reactive: () => I,
  ref: () => E,
  retry: () => kt,
  serializeForm: () => dt,
  sessionStorage: () => Fe,
  set: () => Ot,
  setPrefix: () => ut,
  sleep: () => Je,
  snakeCase: () => Jt,
  stop: () => Ne,
  stringifyQuery: () => Dt,
  throttle: () => St,
  toRaw: () => Ve,
  track: () => Q,
  trigger: () => Z,
  uniqueId: () => Mt,
  unref: () => Ke,
  unset: () => At,
  useCookie: () => wt,
  useDebouncedValidation: () => pt,
  useFetch: () => $e,
  useFieldArray: () => ht,
  useForm: () => Ue,
  useFormSubmit: () => ft,
  useLocalStorage: () => yt,
  usePolling: () => st,
  useSessionStorage: () => gt,
  useStorage: () => Ae,
  useWebSocket: () => Pe,
  uuid: () => Rt,
  watch: () => fe
});
module.exports = __toCommonJS(monkeysjs_esm_exports);
let F = null;
const se = [], me = /* @__PURE__ */ new WeakMap();
function X(e, n = {}) {
  const t = () => {
    ke(t), F = t, se.push(t);
    try {
      return e();
    } finally {
      se.pop(), F = se[se.length - 1];
    }
  };
  return t.deps = /* @__PURE__ */ new Set(), t.options = n, t.active = true, n.lazy || t(), t;
}
function ke(e) {
  e.deps.forEach((n) => {
    n.delete(e);
  }), e.deps.clear();
}
function Ne(e) {
  e.active && (ke(e), e.active = false, e.options.onStop && e.options.onStop());
}
function Q(e, n) {
  if (!F) return;
  let t = me.get(e);
  t || me.set(e, t = /* @__PURE__ */ new Map());
  let r = t.get(n);
  r || t.set(n, r = /* @__PURE__ */ new Set()), r.has(F) || (r.add(F), F.deps.add(r));
}
function Z(e, n, t = "set") {
  const r = me.get(e);
  if (!r) return;
  const s = /* @__PURE__ */ new Set(), o = /* @__PURE__ */ new Set(), i = (a) => {
    a && a.forEach((c) => {
      c !== F && (c.options.computed ? o.add(c) : s.add(c));
    });
  };
  i(r.get(n)), t === "add" && Array.isArray(e) && i(r.get("length")), n === "length" && Array.isArray(e) && r.forEach((a, c) => {
    c >= e.length && i(a);
  }), o.forEach((a) => {
    a.options.scheduler ? a.options.scheduler(a) : a();
  }), s.forEach((a) => {
    a.options.scheduler ? a.options.scheduler(a) : a();
  });
}
const He = {
  get(e, n, t) {
    if (n === "__isReactive") return true;
    if (n === "__raw") return e;
    const r = Reflect.get(e, n, t);
    return Q(e, n), r !== null && typeof r == "object" ? I(r) : r;
  },
  set(e, n, t, r) {
    const s = e[n], o = Array.isArray(e) ? Number(n) < e.length : Object.prototype.hasOwnProperty.call(e, n), i = Reflect.set(e, n, t, r);
    return o ? t !== s && (t === t || s === s) && Z(e, n, "set") : Z(e, n, "add"), i;
  },
  deleteProperty(e, n) {
    const t = Object.prototype.hasOwnProperty.call(e, n), r = Reflect.deleteProperty(e, n);
    return t && r && Z(e, n, "delete"), r;
  },
  has(e, n) {
    return Q(e, n), Reflect.has(e, n);
  },
  ownKeys(e) {
    return Q(e, Array.isArray(e) ? "length" : /* @__PURE__ */ Symbol.for("iterate")), Reflect.ownKeys(e);
  }
}, Ce = /* @__PURE__ */ new WeakMap();
function I(e) {
  if (e === null || typeof e != "object" || e.__isReactive)
    return e;
  const n = Ce.get(e);
  if (n)
    return n;
  const t = new Proxy(e, He);
  return Ce.set(e, t), t;
}
function E(e) {
  const n = {
    __isRef: true,
    get value() {
      return Q(n, "value"), e;
    },
    set value(t) {
      t !== e && (e = t, Z(n, "value"));
    }
  };
  return n;
}
function Ke(e) {
  return e != null && e.__isRef ? e.value : e;
}
function ye(e) {
  return (e == null ? void 0 : e.__isRef) === true;
}
function ge(e) {
  return (e == null ? void 0 : e.__isReactive) === true;
}
function Ve(e) {
  return (e == null ? void 0 : e.__raw) || e;
}
function ae(e) {
  let n, t;
  typeof e == "function" ? (n = e, t = () => {
    console.warn("Computed property is readonly");
  }) : (n = e.get, t = e.set);
  let r, s = true;
  const o = X(n, {
    lazy: true,
    computed: true,
    scheduler: () => {
      s || (s = true, Z(i, "value"));
    }
  }), i = {
    __isRef: true,
    __isComputed: true,
    get value() {
      return s && (r = o(), s = false), Q(i, "value"), r;
    },
    set value(a) {
      t(a);
    }
  };
  return i;
}
function fe(e, n, t = {}) {
  let r, s, o;
  const i = (u) => {
    o = u;
  };
  typeof e == "function" ? r = e : ye(e) ? r = () => e.value : ge(e) ? r = () => le(e) : Array.isArray(e) ? r = () => e.map((u) => ye(u) ? u.value : ge(u) ? le(u) : typeof u == "function" ? u() : u) : r = () => e;
  const a = () => {
    o && o();
    const u = c();
    (t.deep || u !== s || typeof u == "object" && u !== null) && (n(u, s, i), s = u);
  }, c = X(r, {
    lazy: true,
    scheduler: () => {
      t.flush === "sync" ? a() : t.flush === "post" ? Promise.resolve().then(() => queueMicrotask(a)) : queueMicrotask(a);
    }
  });
  return t.immediate ? a() : s = c(), () => {
    Ne(c), o && o();
  };
}
function le(e, n = /* @__PURE__ */ new Set()) {
  return e === null || typeof e != "object" || n.has(e) || (n.add(e), Array.isArray(e) ? e.forEach((t) => le(t, n)) : Object.keys(e).forEach((t) => le(e[t], n))), e;
}
const Te = /* @__PURE__ */ new Set();
function Qe(e) {
  try {
    e();
  } finally {
    Te.forEach((n) => n()), Te.clear();
  }
}
const Y = {
  reactive: I,
  ref: E,
  unref: Ke,
  isRef: ye,
  isReactive: ge,
  toRaw: Ve,
  computed: ae,
  watch: fe,
  effect: X,
  stop: Ne,
  batch: Qe,
  track: Q,
  trigger: Z
}, K = {
  IDLE: "idle",
  LOADING: "loading",
  SUCCESS: "success",
  ERROR: "error"
}, we = {
  baseURL: "",
  timeout: 3e4,
  headers: {
    "Content-Type": "application/json"
  },
  retries: 0,
  retryDelay: 1e3,
  retryBackoff: "exponential",
  // 'linear' | 'exponential'
  retryCondition: (e) => e.status >= 500 || e.status === 0,
  cache: false,
  cacheTTL: 5 * 60 * 1e3,
  // 5 minutes
  dedupeRequests: true,
  credentials: "same-origin",
  responseType: "json"
  // 'json' | 'text' | 'blob' | 'arrayBuffer'
}, ee = [], te = [], z = /* @__PURE__ */ new Map(), ne = /* @__PURE__ */ new Map(), oe = /* @__PURE__ */ new Map();
class ie extends Error {
  constructor(n, t, r, s) {
    super(n), this.name = "HttpError", this.status = t, this.response = r, this.request = s;
  }
}
function Ze(e) {
  const { method: n, url: t, params: r, data: s } = e;
  return `${n}:${t}:${JSON.stringify(r || {})}:${JSON.stringify(s || {})}`;
}
function Ie(e, n) {
  const t = ne.get(e);
  return t ? Date.now() - t < n : false;
}
function Xe(e, n, t) {
  let r = n.startsWith("http") ? n : `${e}${n}`;
  if (t && Object.keys(t).length > 0) {
    const s = new URLSearchParams();
    Object.entries(t).forEach(([i, a]) => {
      a != null && (Array.isArray(a) ? a.forEach((c) => s.append(i, c)) : s.append(i, a));
    });
    const o = r.includes("?") ? "&" : "?";
    r += o + s.toString();
  }
  return r;
}
function Ye(e) {
  return new Promise((n) => setTimeout(n, e));
}
function et(e, n, t) {
  return t === "exponential" ? n * Math.pow(2, e) : n * (e + 1);
}
async function tt(e) {
  let n = { ...e };
  for (const t of ee)
    try {
      n = await t.fulfilled(n);
    } catch (r) {
      if (t.rejected)
        n = await t.rejected(r);
      else
        throw r;
    }
  return n;
}
async function nt(e, n) {
  let t = e;
  for (const r of te)
    try {
      t = await r.fulfilled(t, n);
    } catch (s) {
      if (r.rejected)
        t = await r.rejected(s, n);
      else
        throw s;
    }
  return t;
}
async function Le(e, n) {
  switch (n) {
    case "json":
      const t = await e.text();
      try {
        return t ? JSON.parse(t) : null;
      } catch (e2) {
        return t;
      }
    case "text":
      return e.text();
    case "blob":
      return e.blob();
    case "arrayBuffer":
      return e.arrayBuffer();
    case "formData":
      return e.formData();
    default:
      return e.json();
  }
}
async function rt(e) {
  const n = { ...we, ...e }, {
    baseURL: t,
    url: r,
    method: s = "GET",
    headers: o,
    data: i,
    params: a,
    timeout: c,
    retries: u,
    retryDelay: f,
    retryBackoff: p,
    retryCondition: h,
    cache: b,
    cacheTTL: $,
    dedupeRequests: S,
    credentials: U,
    responseType: W,
    onUploadProgress: D,
    onDownloadProgress: _,
    signal: R
  } = n, M = Xe(t, r, a), L = Ze({ method: s, url: M, params: a, data: i });
  if (b && s === "GET" && Ie(L, $))
    return z.get(L);
  if (S && s === "GET" && oe.has(L))
    return oe.get(L);
  const A = new AbortController(), G = c > 0 ? setTimeout(() => A.abort(), c) : null, x = (R && R.aborted, A.signal);
  R && R.addEventListener("abort", () => A.abort());
  const C = {
    method: s.toUpperCase(),
    headers: { ...we.headers, ...o },
    credentials: U,
    signal: x
  };
  i && s !== "GET" && (i instanceof FormData ? (delete C.headers["Content-Type"], C.body = i) : typeof i == "object" ? C.body = JSON.stringify(i) : C.body = i);
  const q = (async () => {
    let g;
    for (let y = 0; y <= u; y++)
      try {
        const m = await fetch(M, C);
        if (!m.ok) {
          const d = await Le(m.clone(), W);
          throw new ie(
            `HTTP ${m.status}: ${m.statusText}`,
            m.status,
            d,
            { url: M, method: s, data: i }
          );
        }
        const l = {
          data: await Le(m, W),
          status: m.status,
          statusText: m.statusText,
          headers: Object.fromEntries(m.headers.entries()),
          config: n
        };
        return b && s === "GET" && (z.set(L, l), ne.set(L, Date.now())), l;
      } catch (m) {
        if (g = m, m.name === "AbortError")
          throw new ie("Request aborted", 0, null, { url: M, method: s, data: i });
        if (y < u && (m instanceof ie ? h(m) : true)) {
          const l = et(y, f, p);
          await Ye(l);
          continue;
        }
        throw m;
      } finally {
        G && clearTimeout(G);
      }
    throw g;
  })();
  return S && s === "GET" && (oe.set(L, q), q.finally(() => oe.delete(L)).catch(() => {
  })), q;
}
async function P(e) {
  const n = await tt(e), t = await rt(n);
  return nt(t, n);
}
function xe(e = {}) {
  const n = { ...we, ...e };
  return {
    // Configuration
    defaults: n,
    // HTTP methods
    get(r, s = {}) {
      return P({ ...n, ...s, url: r, method: "GET" });
    },
    post(r, s, o = {}) {
      return P({ ...n, ...o, url: r, method: "POST", data: s });
    },
    put(r, s, o = {}) {
      return P({ ...n, ...o, url: r, method: "PUT", data: s });
    },
    patch(r, s, o = {}) {
      return P({ ...n, ...o, url: r, method: "PATCH", data: s });
    },
    delete(r, s = {}) {
      return P({ ...n, ...s, url: r, method: "DELETE" });
    },
    head(r, s = {}) {
      return P({ ...n, ...s, url: r, method: "HEAD" });
    },
    options(r, s = {}) {
      return P({ ...n, ...s, url: r, method: "OPTIONS" });
    },
    // Generic request
    request(r) {
      return P({ ...n, ...r });
    },
    // Interceptors
    interceptors: {
      request: {
        use(r, s) {
          const o = ee.length;
          return ee.push({ fulfilled: r, rejected: s }), o;
        },
        eject(r) {
          ee.splice(r, 1);
        },
        clear() {
          ee.length = 0;
        }
      },
      response: {
        use(r, s) {
          const o = te.length;
          return te.push({ fulfilled: r, rejected: s }), o;
        },
        eject(r) {
          te.splice(r, 1);
        },
        clear() {
          te.length = 0;
        }
      }
    },
    // Cache management
    cache: {
      get(r) {
        return z.get(r);
      },
      set(r, s, o = n.cacheTTL) {
        z.set(r, s), ne.set(r, Date.now());
      },
      delete(r) {
        z.delete(r), ne.delete(r);
      },
      clear() {
        z.clear(), ne.clear();
      },
      has(r) {
        return z.has(r) && Ie(r, n.cacheTTL);
      }
    }
  };
}
function $e(e, n = {}) {
  const t = E(null), r = E(null), s = E(K.IDLE), o = E(false), i = E(false), a = E(false), c = E(null), u = I({
    url: e,
    ...n
  }), f = E(null);
  async function p($ = {}) {
    f.value && f.value.abort(), f.value = new AbortController(), s.value = K.LOADING, o.value = true, i.value = false, a.value = false, r.value = null;
    try {
      const S = await P({
        ...u,
        ...$,
        url: $.url || u.url,
        signal: f.value.signal
      });
      return t.value = S.data, c.value = S, s.value = K.SUCCESS, a.value = true, S;
    } catch (S) {
      if (r.value = S, s.value = K.ERROR, i.value = true, n.throwOnError)
        throw S;
      return null;
    } finally {
      o.value = false;
    }
  }
  function h() {
    f.value && f.value.abort();
  }
  function b() {
    t.value = null, r.value = null, s.value = K.IDLE, o.value = false, i.value = false, a.value = false, c.value = null;
  }
  return n.immediate !== false && p(), {
    data: t,
    error: r,
    status: s,
    isLoading: o,
    isError: i,
    isSuccess: a,
    response: c,
    execute: p,
    abort: h,
    reset: b,
    refresh: p
  };
}
function st(e, n = 5e3, t = {}) {
  const r = $e(e, { ...t, immediate: false }), s = E(false);
  let o = null;
  async function i() {
    s.value && (await r.execute(), s.value && (o = setTimeout(i, n)));
  }
  function a() {
    s.value || (s.value = true, i());
  }
  function c() {
    s.value = false, o && (clearTimeout(o), o = null);
  }
  function u() {
    s.value ? c() : a();
  }
  return {
    ...r,
    isPolling: s,
    start: a,
    stop: c,
    toggle: u
  };
}
const ot = xe(), he = {
  createClient: xe,
  useFetch: $e,
  usePolling: st,
  http: ot,
  HttpError: ie,
  RequestState: K
}, N = {
  CONNECTING: "connecting",
  OPEN: "open",
  CLOSING: "closing",
  CLOSED: "closed"
}, at = {
  reconnect: true,
  reconnectAttempts: 1 / 0,
  reconnectDelay: 1e3,
  reconnectDelayMax: 3e4,
  reconnectBackoff: "exponential",
  heartbeat: false,
  heartbeatInterval: 3e4,
  heartbeatMessage: "ping",
  heartbeatTimeout: 1e4,
  protocols: [],
  immediate: true
};
function Pe(e, n = {}) {
  const t = { ...at, ...n }, r = E(N.CLOSED), s = E(null), o = E(null), i = E(null), a = E(null), c = E(0), u = E(null), f = E(null), p = E(null), h = E(null), b = E(false), $ = E([]), S = {
    open: /* @__PURE__ */ new Set(),
    message: /* @__PURE__ */ new Set(),
    error: /* @__PURE__ */ new Set(),
    close: /* @__PURE__ */ new Set()
  };
  function U() {
    const { reconnectDelay: g, reconnectDelayMax: y, reconnectBackoff: m } = t;
    let O;
    return m === "exponential" ? O = g * Math.pow(2, c.value) : O = g * (c.value + 1), Math.min(O, y);
  }
  function W() {
    t.heartbeat && (D(), f.value = setInterval(() => {
      var g;
      ((g = u.value) == null ? void 0 : g.readyState) === WebSocket.OPEN && (A(t.heartbeatMessage), p.value = setTimeout(() => {
        var y;
        console.warn("WebSocket heartbeat timeout"), (y = u.value) == null || y.close();
      }, t.heartbeatTimeout));
    }, t.heartbeatInterval));
  }
  function D() {
    f.value && (clearInterval(f.value), f.value = null), p.value && (clearTimeout(p.value), p.value = null);
  }
  function _() {
    p.value && (clearTimeout(p.value), p.value = null);
  }
  function R() {
    var g;
    if (((g = u.value) == null ? void 0 : g.readyState) !== WebSocket.OPEN) {
      b.value = false, r.value = N.CONNECTING;
      try {
        u.value = new WebSocket(e, t.protocols), u.value.onopen = (y) => {
          for (r.value = N.OPEN, o.value = null, c.value = 0; $.value.length > 0; ) {
            const m = $.value.shift();
            L(m);
          }
          W(), S.open.forEach((m) => m(y));
        }, u.value.onmessage = (y) => {
          _();
          let m = y.data;
          try {
            m = JSON.parse(y.data);
          } catch (e2) {
          }
          s.value = m, i.value = y.data, a.value = Date.now(), S.message.forEach((O) => O(m, y));
        }, u.value.onerror = (y) => {
          o.value = y, S.error.forEach((m) => m(y));
        }, u.value.onclose = (y) => {
          if (r.value = N.CLOSED, D(), S.close.forEach((m) => m(y)), !b.value && t.reconnect && c.value < t.reconnectAttempts) {
            const m = U();
            c.value++, h.value = setTimeout(() => {
              R();
            }, m);
          }
        };
      } catch (y) {
        o.value = y, r.value = N.CLOSED;
      }
    }
  }
  function M(g = 1e3, y = "") {
    b.value = true, r.value = N.CLOSING, h.value && (clearTimeout(h.value), h.value = null), D(), u.value && (u.value.close(g, y), u.value = null), r.value = N.CLOSED;
  }
  function L(g) {
    var y;
    if (((y = u.value) == null ? void 0 : y.readyState) === WebSocket.OPEN) {
      const m = typeof g == "object" ? JSON.stringify(g) : g;
      return u.value.send(m), true;
    }
    return false;
  }
  function A(g, y = {}) {
    var O;
    const { queue: m = true } = y;
    return ((O = u.value) == null ? void 0 : O.readyState) === WebSocket.OPEN ? L(g) : m ? ($.value.push(g), true) : false;
  }
  function G(g, y = {}) {
    const { timeout: m = 1e4, matcher: O } = y;
    return new Promise((l, d) => {
      const w = setTimeout(() => {
        C("message", B), d(new Error("WebSocket response timeout"));
      }, m), B = (k, H) => {
        (O ? O(k, g) : true) && (clearTimeout(w), C("message", B), l(k));
      };
      x("message", B), A(g) || (clearTimeout(w), C("message", B), d(new Error("Failed to send message")));
    });
  }
  function x(g, y) {
    return S[g] && S[g].add(y), () => C(g, y);
  }
  function C(g, y) {
    S[g] && S[g].delete(y);
  }
  function q(g, y) {
    const m = (...O) => {
      C(g, m), y(...O);
    };
    return x(g, m), () => C(g, m);
  }
  return t.immediate && R(), {
    // Refs
    status: r,
    data: s,
    error: o,
    lastMessage: i,
    lastMessageTime: a,
    reconnectCount: c,
    // Computed (as getters for convenience, but users should prefer refs)
    get isConnected() {
      return r.value === N.OPEN;
    },
    get isConnecting() {
      return r.value === N.CONNECTING;
    },
    // Methods
    open: R,
    close: M,
    send: A,
    sendAsync: G,
    on: x,
    off: C,
    once: q,
    // WebSocket instance
    ws: u
  };
}
function it(e, n = {}) {
  const t = /* @__PURE__ */ new Map(), r = {
    open: /* @__PURE__ */ new Set(),
    message: /* @__PURE__ */ new Set(),
    error: /* @__PURE__ */ new Set(),
    close: /* @__PURE__ */ new Set()
  };
  function s(f = "", p = {}) {
    const h = f ? `${e}${f}` : e, b = h;
    if (t.has(b))
      return t.get(b);
    const $ = Pe(h, { ...n, ...p });
    return Object.keys(r).forEach((S) => {
      r[S].forEach((U) => {
        $.on(S, U);
      });
    }), t.set(b, $), $;
  }
  function o(f = "") {
    const p = f ? `${e}${f}` : e, h = t.get(p);
    h && (h.close(), t.delete(p));
  }
  function i() {
    t.forEach((f) => f.close()), t.clear();
  }
  function a(f, p) {
    var h;
    return (h = r[f]) == null || h.add(p), t.forEach((b) => {
      b.on(f, p);
    }), () => c(f, p);
  }
  function c(f, p) {
    var h;
    (h = r[f]) == null || h.delete(p), t.forEach((b) => {
      b.off(f, p);
    });
  }
  function u(f) {
    t.forEach((p) => {
      p.send(f);
    });
  }
  return {
    connect: s,
    disconnect: o,
    disconnectAll: i,
    on: a,
    off: c,
    broadcast: u,
    connections: t
  };
}
const ct = {
  useWebSocket: Pe,
  createWebSocketClient: it,
  WebSocketState: N
}, re = /* @__PURE__ */ new Map(), je = /* @__PURE__ */ new Map();
let j = "m-";
const be = /* @__PURE__ */ new Map();
function ut(e) {
  j = e;
}
function v(e, n) {
  re.set(e, n);
}
function lt(e, n) {
  je.set(e, n);
}
function T(e, n, t) {
  try {
    const r = (o, i = {}) => {
      const a = new CustomEvent(o, {
        detail: i,
        bubbles: true,
        cancelable: true
      });
      t.dispatchEvent(a);
    };
    return new Function(
      "$data",
      "$el",
      "$refs",
      "$event",
      "$dispatch",
      "$watch",
      `with($data) { return ${e} }`
    )(
      n.$data,
      t,
      n.$refs,
      n.$event,
      r,
      n.$watch
    );
  } catch (r) {
    console.warn(`MonkeysJS: Error evaluating "${e}":`, r);
    return;
  }
}
function J(e, n, t, r) {
  try {
    const s = (i, a = {}) => {
      const c = new CustomEvent(i, {
        detail: a,
        bubbles: true,
        cancelable: true
      });
      t.dispatchEvent(c);
    };
    return new Function(
      "$data",
      "$el",
      "$refs",
      "$event",
      "$dispatch",
      "$watch",
      `with($data) { ${e} }`
    )(
      n.$data,
      t,
      n.$refs,
      r,
      s,
      n.$watch
    );
  } catch (s) {
    console.warn(`MonkeysJS: Error executing "${e}":`, s);
  }
}
v("data", {
  init(e, n, t) {
    const r = n ? T(n, t, e) : {};
    t.$data = I({ ...t.$data, ...r });
  }
});
v("text", {
  effect(e, n, t) {
    const r = T(n, t, e);
    e.textContent = r != null ? r : "";
  }
});
v("html", {
  effect(e, n, t) {
    const r = T(n, t, e);
    e.innerHTML = r != null ? r : "";
  }
});
v("show", {
  effect(e, n, t) {
    const r = T(n, t, e);
    e.style.display = r ? "" : "none";
  }
});
v("if", {
  init(e, n, t) {
    const r = document.createComment("m-if"), s = e.cloneNode(true);
    e.parentNode.insertBefore(r, e), e.remove(), t._ifState = { placeholder: r, template: s, currentEl: null };
  },
  effect(e, n, t) {
    const r = T(n, t, e), { placeholder: s, template: o, currentEl: i } = t._ifState;
    if (r && !i) {
      const a = o.cloneNode(true);
      a.removeAttribute(`${j}if`), s.parentNode.insertBefore(a, s.nextSibling), t._ifState.currentEl = a, de(a, t);
    } else !r && i && (i.remove(), t._ifState.currentEl = null);
  }
});
v("for", {
  init(e, n, t) {
    const r = document.createComment("m-for"), s = e.cloneNode(true);
    s.removeAttribute(`${j}for`), e.parentNode.insertBefore(r, e), e.remove();
    const o = n.match(/^\s*(?:\(([^,]+),\s*([^)]+)\)|([^\s]+))\s+(?:in|of)\s+(.+)\s*$/);
    if (!o) {
      console.warn(`MonkeysJS: Invalid m-for expression: ${n}`);
      return;
    }
    const i = o[3] || o[1], a = o[2] || "index", c = o[4];
    t._forState = {
      placeholder: r,
      template: s,
      elements: [],
      itemName: i,
      indexName: a,
      listExpression: c
    };
  },
  effect(e, n, t) {
    const { placeholder: r, template: s, elements: o, itemName: i, indexName: a, listExpression: c } = t._forState, u = T(c, t, e) || [];
    o.forEach((f) => f.remove()), o.length = 0, u.forEach((f, p) => {
      const h = s.cloneNode(true), b = De({
        ...t.$data,
        [i]: f,
        [a]: p
      }, t.$refs);
      r.parentNode.insertBefore(h, r), o.push(h), de(h, b);
    });
  }
});
v("bind", {
  effect(e, n, t, r, s) {
    if (s) {
      const o = T(n, t, e);
      Me(e, s, o);
    } else {
      const o = T(n, t, e);
      o && typeof o == "object" && Object.entries(o).forEach(([i, a]) => {
        Me(e, i, a);
      });
    }
  }
});
v("on", {
  init(e, n, t, r, s) {
    if (!s) return;
    const o = (a) => {
      if (r.includes("prevent") && a.preventDefault(), r.includes("stop") && a.stopPropagation(), !(r.includes("self") && a.target !== e)) {
        if (r.includes("once") && e.removeEventListener(s, o), a instanceof KeyboardEvent) {
          const c = ["enter", "tab", "delete", "esc", "space", "up", "down", "left", "right"], u = a.key.toLowerCase();
          for (const f of r)
            if (c.includes(f) && u !== {
              enter: "enter",
              tab: "tab",
              delete: "delete",
              esc: "escape",
              space: " ",
              up: "arrowup",
              down: "arrowdown",
              left: "arrowleft",
              right: "arrowright"
            }[f])
              return;
        }
        t.$event = a, J(n, t, e, a);
      }
    }, i = {
      capture: r.includes("capture"),
      passive: r.includes("passive"),
      once: r.includes("once")
    };
    r.includes("window") ? window.addEventListener(s, o, i) : r.includes("document") ? document.addEventListener(s, o, i) : e.addEventListener(s, o, i);
  }
});
v("model", {
  init(e, n, t, r) {
    const s = e.type === "checkbox", o = e.type === "radio", a = e.tagName === "SELECT" && e.multiple, c = r.includes("lazy") ? "change" : "input";
    e.addEventListener(c, (u) => {
      let f;
      if (s) {
        const p = T(n, t, e);
        Array.isArray(p) ? f = e.checked ? [...p, e.value] : p.filter((h) => h !== e.value) : f = e.checked;
      } else o ? f = e.value : a ? f = Array.from(e.selectedOptions).map((p) => p.value) : (f = e.value, r.includes("number") ? f = parseFloat(f) || 0 : r.includes("trim") && (f = f.trim()));
      try {
        new Function("$data", "value", `with($data) { ${n} = value }`)(t.$data, f);
      } catch (p) {
        console.warn(`MonkeysJS: Error setting model "${n}":`, p);
      }
    });
  },
  effect(e, n, t, r) {
    const s = T(n, t, e), o = e.type === "checkbox", i = e.type === "radio", a = e.tagName === "SELECT";
    o ? Array.isArray(s) ? e.checked = s.includes(e.value) : e.checked = !!s : i ? e.checked = e.value === s : a ? requestAnimationFrame(() => {
      e.multiple && Array.isArray(s) ? Array.from(e.options).forEach((c) => {
        c.selected = s.includes(c.value);
      }) : e.value = s != null ? s : "";
    }) : e.value = s != null ? s : "";
  }
});
v("ref", {
  init(e, n, t) {
    t.$refs[n] = e;
  }
});
v("cloak", {
  init(e) {
    e.removeAttribute(`${j}cloak`);
  }
});
v("fetch", {
  init(e, n, t, r) {
    var a;
    const s = T(n, t, e), o = ((a = r[0]) == null ? void 0 : a.toUpperCase()) || "GET", i = $e(s, { method: o, immediate: false });
    t.$fetch = i, t.$data.$loading = i.isLoading, t.$data.$error = i.error, t.$data.$data = i.data;
  },
  effect(e, n, t) {
    const r = T(n, t, e);
    r && t.$fetch && t.$fetch.execute({ url: r });
  }
});
v("class", {
  effect(e, n, t) {
    const r = T(n, t, e);
    typeof r == "string" ? e.className = r : Array.isArray(r) ? e.className = r.filter(Boolean).join(" ") : typeof r == "object" && Object.entries(r).forEach(([s, o]) => {
      e.classList.toggle(s, !!o);
    });
  }
});
v("style", {
  effect(e, n, t) {
    const r = T(n, t, e);
    typeof r == "string" ? e.style.cssText = r : typeof r == "object" && Object.entries(r).forEach(([s, o]) => {
      s.startsWith("--") ? e.style.setProperty(s, o) : e.style[s] = o;
    });
  }
});
v("transition", {
  init(e, n, t) {
    const r = n || "fade";
    e.dataset.transition = r, t._transition = {
      enter: `${r}-enter`,
      enterActive: `${r}-enter-active`,
      enterTo: `${r}-enter-to`,
      leave: `${r}-leave`,
      leaveActive: `${r}-leave-active`,
      leaveTo: `${r}-leave-to`
    };
  }
});
v("teleport", {
  init(e, n, t) {
    const r = document.querySelector(n);
    if (!r) {
      console.warn(`MonkeysJS: Teleport target "${n}" not found`);
      return;
    }
    const s = document.createComment("m-teleport");
    e.parentNode.insertBefore(s, e), r.appendChild(e), t._teleportState = { placeholder: s, target: r, originalParent: s.parentNode };
  }
});
v("lazy", {
  init(e, n, t) {
    const r = new IntersectionObserver((s) => {
      s.forEach((o) => {
        o.isIntersecting && (t.$data.$lazy = true, e.classList.add("m-lazy-loaded"), r.unobserve(e), n && J(n, t, e));
      });
    }, { threshold: 0.1 });
    t.$data.$lazy = false, r.observe(e), t._lazyObserver = r;
  }
});
v("intersect", {
  init(e, n, t, r) {
    const s = r.includes("once"), o = r.find((a) => !isNaN(a)) || 0, i = new IntersectionObserver((a) => {
      a.forEach((c) => {
        t.$data.$visible = c.isIntersecting, t.$data.$intersectionRatio = c.intersectionRatio, c.isIntersecting && (J(n, t, e), s && i.unobserve(e));
      });
    }, { threshold: parseFloat(o) });
    i.observe(e);
  }
});
v("debounce", {
  init(e, n, t, r, s) {
    const o = parseInt(r[0]) || 300;
    let i = null;
    e.addEventListener(s || "input", (a) => {
      i && clearTimeout(i), i = setTimeout(() => {
        t.$event = a, J(n, t, e, a);
      }, o);
    });
  }
});
v("throttle", {
  init(e, n, t, r, s) {
    const o = parseInt(r[0]) || 300;
    let i = false;
    e.addEventListener(s || "scroll", (a) => {
      i || (t.$event = a, J(n, t, e, a), i = true, setTimeout(() => i = false, o));
    });
  }
});
v("clipboard", {
  init(e, n, t, r) {
    e.addEventListener("click", async () => {
      const s = T(n, t, e);
      try {
        await navigator.clipboard.writeText(s), e.classList.add("m-clipboard-success"), t.$data.$copied = true, setTimeout(() => {
          e.classList.remove("m-clipboard-success"), t.$data.$copied = false;
        }, 2e3);
      } catch (o) {
        console.error("Failed to copy:", o), t.$data.$copied = false;
      }
    });
  }
});
v("portal", re.get("teleport"));
v("animate", {
  init(e, n, t, r) {
    const s = n || "animate", o = r[0] || "load", i = () => {
      e.classList.add(s), e.addEventListener("animationend", () => {
        r.includes("keep") || e.classList.remove(s);
      }, { once: true });
    };
    switch (o) {
      case "load":
        i();
        break;
      case "hover":
        e.addEventListener("mouseenter", i);
        break;
      case "click":
        e.addEventListener("click", i);
        break;
      case "visible":
        const a = new IntersectionObserver((c) => {
          c[0].isIntersecting && (i(), a.unobserve(e));
        });
        a.observe(e);
        break;
    }
  }
});
v("hotkey", {
  init(e, n, t, r) {
    const s = r.join("+").toLowerCase(), o = (i) => {
      const a = [];
      (i.ctrlKey || i.metaKey) && a.push("ctrl"), i.shiftKey && a.push("shift"), i.altKey && a.push("alt"), a.push(i.key.toLowerCase()), a.join("+") === s && (i.preventDefault(), J(n, t, e, i));
    };
    document.addEventListener("keydown", o), t._hotkeyHandler = o;
  }
});
v("persist", {
  init(e, n, t, r) {
    const s = n || "m-persist-" + Math.random().toString(36).slice(2), o = r.includes("session") ? sessionStorage : localStorage;
    try {
      const c = o.getItem(s);
      if (c) {
        const u = JSON.parse(c);
        Object.assign(t.$data, u);
      }
    } catch (e2) {
    }
    const i = r.filter((c) => c !== "session" && c !== "local"), a = i.length > 0 ? i : Object.keys(t.$data);
    X(() => {
      const c = {};
      a.forEach((u) => {
        u in t.$data && (c[u] = t.$data[u]);
      }), o.setItem(s, JSON.stringify(c));
    });
  }
});
function Me(e, n, t) {
  n === "class" ? typeof t == "object" && !Array.isArray(t) ? Object.entries(t).forEach(([r, s]) => {
    e.classList.toggle(r, !!s);
  }) : e.className = Array.isArray(t) ? t.join(" ") : t : n === "style" ? typeof t == "object" ? Object.assign(e.style, t) : e.style.cssText = t : n.startsWith("data-") ? e.dataset[n.slice(5)] = t : t === true ? e.setAttribute(n, "") : t === false || t === null || t === void 0 ? e.removeAttribute(n) : e.setAttribute(n, t);
}
v("mask", {
  init(e, n, t) {
    if (e.tagName !== "INPUT") return;
    const r = n, s = (o) => {
      let i = 0, a = "";
      for (let c = 0; c < r.length; c++) {
        const u = r[c];
        if (!["9", "a", "*"].includes(u)) {
          a += u, o[i] === u && i++;
          continue;
        }
        let f = false;
        for (; i < o.length; ) {
          const p = o[i];
          if (i++, u === "9" && /\d/.test(p) || u === "a" && /[a-zA-Z]/.test(p) || u === "*" && /[a-zA-Z0-9]/.test(p)) {
            a += p, f = true;
            break;
          }
        }
        if (!f) break;
      }
      return a;
    };
    e.addEventListener("input", (o) => {
      const i = o.target, a = i.value, c = s(a);
      c !== a && (i.value = c, i.dispatchEvent(new Event("input", { bubbles: true })));
    });
  }
});
v("effect", {
  effect(e, n, t) {
    J(n, t, e);
  }
});
function De(e = {}, n = {}) {
  const t = {
    $data: I(e),
    $refs: n,
    $event: null,
    $fetch: null
  };
  return t.$dispatch = (r, s = {}) => {
  }, t.$watch = (r, s, o) => fe(r, s, o), t;
}
function Re(e) {
  const n = e.name;
  if (n.startsWith("@"))
    return {
      name: "on",
      arg: n.slice(1).split(".")[0],
      modifiers: n.slice(1).split(".").slice(1),
      expression: e.value
    };
  if (n.startsWith(":"))
    return {
      name: "bind",
      arg: n.slice(1).split(".")[0],
      modifiers: n.slice(1).split(".").slice(1),
      expression: e.value
    };
  if (!n.startsWith(j)) return null;
  const t = n.slice(j.length), [r, ...s] = t.split("."), [o, i] = r.split(":");
  return {
    name: o,
    arg: i,
    modifiers: s,
    expression: e.value
  };
}
function de(e, n) {
  if (e.nodeType !== Node.ELEMENT_NODE) return;
  const t = Array.from(e.attributes), r = [], s = ["if", "for"];
  for (const o of s) {
    const i = `${j}${o}`, a = t.find((c) => c.name === i);
    if (a) {
      const c = Re(a);
      if (c) {
        const u = re.get(c.name);
        if (u) {
          u.init && u.init(e, c.expression, n, c.modifiers, c.arg), u.effect && X(() => {
            u.effect(e, c.expression, n, c.modifiers, c.arg);
          });
          return;
        }
      }
    }
  }
  for (const o of t) {
    const i = Re(o);
    if (!i || s.includes(i.name)) continue;
    const a = re.get(i.name);
    a && (a.init && a.init(e, i.expression, n, i.modifiers, i.arg), a.effect && r.push({ handler: a, directive: i }), e.removeAttribute(o.name));
  }
  for (const { handler: o, directive: i } of r)
    X(() => {
      o.effect(e, i.expression, n, i.modifiers, i.arg);
    });
  Array.from(e.children).forEach((o) => {
    de(o, n);
  });
}
function _e(e = {}) {
  let n = null, t = null;
  const r = {
    // Register directive
    directive(s, o) {
      return re.set(s, o), r;
    },
    // Register component
    component(s, o) {
      return je.set(s, o), r;
    },
    // Use plugin
    use(s, o) {
      return typeof s.install == "function" ? s.install(r, o) : typeof s == "function" && s(r, o), r;
    },
    // Provide value for injection
    provide(s, o) {
      return t ? t.$data[s] = o : e[s] = o, r;
    },
    // Mount app
    mount(s) {
      if (n = typeof s == "string" ? document.querySelector(s) : s, !n)
        return console.error(`MonkeysJS: Could not find element "${s}"`), r;
      const o = n.getAttribute(`${j}data`);
      if (o)
        try {
          const i = new Function(`return ${o}`)();
          Object.assign(e, i);
        } catch (e2) {
        }
      return t = De(e), de(n, t), be.set(n, { app: r, context: t }), r;
    },
    // Unmount app
    unmount() {
      return n && (be.delete(n), n = null, t = null), r;
    },
    // Get reactive data
    get data() {
      return t == null ? void 0 : t.$data;
    },
    // Get refs
    get refs() {
      return t == null ? void 0 : t.$refs;
    }
  };
  return r;
}
function ve() {
  document.querySelectorAll(`[${j}data]`).forEach((e) => {
    be.has(e) || _e().mount(e);
  });
}
typeof document < "u" && (document.readyState === "loading" ? document.addEventListener("DOMContentLoaded", ve) : ve());
const Be = {
  createApp: _e,
  directive: v,
  component: lt,
  setPrefix: ut,
  autoInit: ve
};
function Ue(e = {}, n = {}) {
  const {
    validateOnChange: t = true,
    validateOnBlur: r = true,
    validateOnSubmit: s = true,
    resetOnSubmit: o = false,
    transform: i,
    validate: a
  } = n, c = I({ ...e }), u = I({}), f = I({}), p = I({}), h = E(false), b = E(0), $ = E(null), S = ae(() => Object.keys(u).length === 0), U = ae(() => Object.values(p).some(Boolean)), W = ae(() => Object.values(f).some(Boolean)), D = {}, _ = {
    required: (l, d, w = "This field is required") => l == null || l === "" || Array.isArray(l) && l.length === 0 ? w : null,
    email: (l, d, w = "Invalid email address") => l ? /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(l) ? null : w : null,
    min: (l, d, w) => l ? typeof l == "number" ? l >= d ? null : w || `Must be at least ${d}` : typeof l == "string" ? l.length >= d ? null : w || `Must be at least ${d} characters` : Array.isArray(l) ? l.length >= d ? null : w || `Must have at least ${d} items` : null : null,
    max: (l, d, w) => l ? typeof l == "number" ? l <= d ? null : w || `Must be at most ${d}` : typeof l == "string" ? l.length <= d ? null : w || `Must be at most ${d} characters` : Array.isArray(l) ? l.length <= d ? null : w || `Must have at most ${d} items` : null : null,
    pattern: (l, d, w = "Invalid format") => l ? (typeof d == "string" ? new RegExp(d) : d).test(l) ? null : w : null,
    url: (l, d, w = "Invalid URL") => {
      if (!l) return null;
      try {
        return new URL(l), null;
      } catch (e2) {
        return w;
      }
    },
    matches: (l, d, w) => l === c[d] ? null : w || `Must match ${d}`,
    custom: (l, d) => d(l, c)
  };
  function R(l) {
    Object.assign(D, l);
  }
  function M(l) {
    const d = D[l];
    if (!d)
      return delete u[l], true;
    const w = c[l], B = Array.isArray(d) ? d : [d];
    for (const k of B) {
      let H = null;
      if (typeof k == "function")
        H = k(w, c);
      else if (typeof k == "object") {
        const { type: pe, value: We, message: Ge } = k;
        _[pe] && (H = _[pe](w, We, Ge));
      } else typeof k == "string" && _[k] && (H = _[k](w));
      if (H)
        return u[l] = H, false;
    }
    return delete u[l], true;
  }
  function L() {
    let l = true;
    if (Object.keys(u).forEach((d) => delete u[d]), a) {
      const d = a(c);
      d && typeof d == "object" && (Object.assign(u, d), l = Object.keys(d).length === 0);
    }
    return Object.keys(D).forEach((d) => {
      M(d) || (l = false);
    }), l;
  }
  function A(l, d) {
    c[l], c[l] = d, p[l] = d !== e[l], t && M(l);
  }
  function G(l) {
    Object.entries(l).forEach(([d, w]) => {
      A(d, w);
    });
  }
  function x(l, d = true) {
    f[l] = d, r && d && M(l);
  }
  function C(l, d) {
    d ? u[l] = d : delete u[l];
  }
  function q(l) {
    delete u[l];
  }
  function g() {
    Object.keys(u).forEach((l) => delete u[l]);
  }
  function y(l) {
    const d = l || e;
    Object.keys(c).forEach((w) => delete c[w]), Object.keys(u).forEach((w) => delete u[w]), Object.keys(f).forEach((w) => delete f[w]), Object.keys(p).forEach((w) => delete p[w]), Object.assign(c, d), l && Object.assign(e, l), $.value = null;
  }
  async function m(l) {
    if (b.value++, $.value = null, s && !L())
      return { success: false, errors: { ...u } };
    h.value = true;
    try {
      const d = i ? i(c) : { ...c }, w = await l(d);
      return o && y(), { success: true, data: w };
    } catch (d) {
      return $.value = d, { success: false, error: d };
    } finally {
      h.value = false;
    }
  }
  function O(l) {
    return {
      get value() {
        return c[l];
      },
      set value(d) {
        A(l, d);
      },
      get error() {
        return u[l];
      },
      get touched() {
        return f[l];
      },
      get dirty() {
        return p[l];
      },
      onChange(d) {
        A(l, d);
      },
      onBlur() {
        x(l);
      },
      // For DOM binding
      bind: {
        value: c[l],
        onInput: (d) => A(l, d.target.value),
        onBlur: () => x(l)
      }
    };
  }
  return {
    // State
    values: c,
    errors: u,
    touched: f,
    dirty: p,
    isSubmitting: h,
    submitCount: b,
    submitError: $,
    // Computed
    isValid: S,
    isDirty: U,
    isTouched: W,
    // Methods
    defineRules: R,
    validate: L,
    validateField: M,
    setValue: A,
    setValues: G,
    setTouched: x,
    setError: C,
    clearError: q,
    clearErrors: g,
    reset: y,
    handleSubmit: m,
    field: O,
    // Validators (for custom rules)
    validators: _
  };
}
function ft(e, n = {}) {
  const {
    method: t = "POST",
    transform: r,
    onSuccess: s,
    onError: o,
    ...i
  } = n;
  return Ue(n.initialValues || {}, {
    ...n
  });
}
function dt(e) {
  const n = new FormData(e), t = {};
  for (const [r, s] of n.entries())
    r in t ? (Array.isArray(t[r]) || (t[r] = [t[r]]), t[r].push(s)) : t[r] = s;
  return t;
}
function pt(e, n = 300) {
  let t = null;
  return (s) => {
    t && clearTimeout(t), t = setTimeout(() => {
      s ? e.validateField(s) : e.validate();
    }, n);
  };
}
function ht(e, n, t = []) {
  const r = I([...t]);
  e.values[n] = r;
  function s(h = {}) {
    r.push(h);
  }
  function o(h = {}) {
    r.unshift(h);
  }
  function i(h, b = {}) {
    r.splice(h, 0, b);
  }
  function a(h) {
    r.splice(h, 1);
  }
  function c(h, b) {
    const $ = r[h];
    r[h] = r[b], r[b] = $;
  }
  function u(h, b) {
    const $ = r.splice(h, 1)[0];
    r.splice(b, 0, $);
  }
  function f(h, b) {
    r[h] = b;
  }
  function p() {
    r.length = 0;
  }
  return {
    fields: r,
    append: s,
    prepend: o,
    insert: i,
    remove: a,
    swap: c,
    move: u,
    replace: f,
    clear: p
  };
}
const mt = {
  useForm: Ue,
  useFormSubmit: ft,
  serializeForm: dt,
  useDebouncedValidation: pt,
  useFieldArray: ht
}, qe = (e) => ({
  get(n) {
    try {
      const t = e.getItem(n);
      return t ? JSON.parse(t) : null;
    } catch (e2) {
      return e.getItem(n);
    }
  },
  set(n, t) {
    try {
      const r = typeof t == "string" ? t : JSON.stringify(t);
      return e.setItem(n, r), true;
    } catch (r) {
      return console.warn("Storage set failed:", r), false;
    }
  },
  remove(n) {
    e.removeItem(n);
  },
  clear() {
    e.clear();
  },
  has(n) {
    return e.getItem(n) !== null;
  },
  keys() {
    return Object.keys(e);
  }
}), ze = typeof window < "u" ? qe(window.localStorage) : Oe(), Fe = typeof window < "u" ? qe(window.sessionStorage) : Oe();
function Oe() {
  const e = /* @__PURE__ */ new Map();
  return {
    get(n) {
      var _a;
      return (_a = e.get(n)) != null ? _a : null;
    },
    set(n, t) {
      return e.set(n, t), true;
    },
    remove(n) {
      e.delete(n);
    },
    clear() {
      e.clear();
    },
    has(n) {
      return e.has(n);
    },
    keys() {
      return Array.from(e.keys());
    }
  };
}
function yt(e, n, t = {}) {
  return Ae(e, n, ze, t);
}
function gt(e, n, t = {}) {
  return Ae(e, n, Fe, t);
}
function Ae(e, n, t, r = {}) {
  const {
    serializer: s = JSON,
    onError: o = console.error,
    listenToStorageChanges: i = true
  } = r, c = E((() => {
    try {
      const u = t.get(e);
      return u !== null ? u : n;
    } catch (u) {
      return o(u), n;
    }
  })());
  return fe(
    () => c.value,
    (u) => {
      try {
        u == null ? t.remove(e) : t.set(e, u);
      } catch (f) {
        o(f);
      }
    },
    { deep: true }
  ), i && typeof window < "u" && window.addEventListener("storage", (u) => {
    var _a;
    if (u.key === e)
      try {
        c.value = u.newValue ? s.parse(u.newValue) : n;
      } catch (e2) {
        c.value = (_a = u.newValue) != null ? _a : n;
      }
  }), c;
}
const ce = {
  get(e) {
    if (typeof document > "u") return null;
    const n = document.cookie.match(
      new RegExp("(?:^|; )" + e.replace(/([.$?*|{}()[\]\\/+^])/g, "\\$1") + "=([^;]*)")
    );
    return n ? decodeURIComponent(n[1]) : null;
  },
  set(e, n, t = {}) {
    if (typeof document > "u") return;
    const {
      path: r = "/",
      domain: s,
      expires: o,
      maxAge: i,
      secure: a,
      sameSite: c = "Lax",
      httpOnly: u
    } = t;
    let f = `${encodeURIComponent(e)}=${encodeURIComponent(n)}`;
    if (r && (f += `; path=${r}`), s && (f += `; domain=${s}`), o) {
      const p = o instanceof Date ? o : new Date(o);
      f += `; expires=${p.toUTCString()}`;
    }
    i !== void 0 && (f += `; max-age=${i}`), a && (f += "; secure"), c && (f += `; samesite=${c}`), u && (f += "; httponly"), document.cookie = f;
  },
  remove(e, n = {}) {
    this.set(e, "", { ...n, maxAge: -1 });
  },
  has(e) {
    return this.get(e) !== null;
  },
  getAll() {
    return typeof document > "u" ? {} : document.cookie.split(";").reduce((e, n) => {
      const [t, r] = n.split("=").map((s) => s.trim());
      return t && (e[decodeURIComponent(t)] = decodeURIComponent(r || "")), e;
    }, {});
  }
};
function wt(e, n, t = {}) {
  var _a;
  const r = E((_a = ce.get(e)) != null ? _a : n);
  return fe(
    () => r.value,
    (s) => {
      s == null ? ce.remove(e, t) : ce.set(e, s, t);
    }
  ), r;
}
function bt(e, n = 1) {
  let t = null, r = null;
  const s = (a = []) => r || (r = new Promise((c, u) => {
    if (typeof indexedDB > "u") {
      u(new Error("IndexedDB not available"));
      return;
    }
    const f = indexedDB.open(e, n);
    f.onerror = () => u(f.error), f.onsuccess = () => {
      t = f.result, c(t);
    }, f.onupgradeneeded = (p) => {
      const h = p.target.result;
      a.forEach((b) => {
        h.objectStoreNames.contains(b.name) || h.createObjectStore(b.name, b.options || { keyPath: "id" });
      });
    };
  }), r), o = (a, c = "readonly") => {
    if (!t) throw new Error("Database not opened");
    return t.transaction(a, c).objectStore(a);
  }, i = (a) => new Promise((c, u) => {
    a.onsuccess = () => c(a.result), a.onerror = () => u(a.error);
  });
  return {
    open: s,
    async get(a, c) {
      await s();
      const u = o(a);
      return i(u.get(c));
    },
    async getAll(a) {
      await s();
      const c = o(a);
      return i(c.getAll());
    },
    async set(a, c) {
      await s();
      const u = o(a, "readwrite");
      return i(u.put(c));
    },
    async add(a, c) {
      await s();
      const u = o(a, "readwrite");
      return i(u.add(c));
    },
    async delete(a, c) {
      await s();
      const u = o(a, "readwrite");
      return i(u.delete(c));
    },
    async clear(a) {
      await s();
      const c = o(a, "readwrite");
      return i(c.clear());
    },
    async count(a) {
      await s();
      const c = o(a);
      return i(c.count());
    },
    close() {
      t && (t.close(), t = null, r = null);
    }
  };
}
const vt = {
  localStorage: ze,
  sessionStorage: Fe,
  createMemoryStorage: Oe,
  useLocalStorage: yt,
  useSessionStorage: gt,
  useStorage: Ae,
  cookies: ce,
  useCookie: wt,
  createIndexedDB: bt
};
function Et(e, n = 300) {
  let t = null;
  const r = (...s) => {
    t && clearTimeout(t), t = setTimeout(() => {
      e.apply(this, s), t = null;
    }, n);
  };
  return r.cancel = () => {
    t && (clearTimeout(t), t = null);
  }, r.flush = (...s) => {
    r.cancel(), e.apply(this, s);
  }, r;
}
function St(e, n = 300) {
  let t = false, r = null;
  const s = (...o) => {
    t ? r = o : (e.apply(this, o), t = true, setTimeout(() => {
      t = false, r && (s.apply(this, r), r = null);
    }, n));
  };
  return s.cancel = () => {
    t = false, r = null;
  }, s;
}
function V(e) {
  if (e === null || typeof e != "object")
    return e;
  if (e instanceof Date)
    return new Date(e.getTime());
  if (e instanceof RegExp)
    return new RegExp(e.source, e.flags);
  if (Array.isArray(e))
    return e.map((t) => V(t));
  if (e instanceof Map) {
    const t = /* @__PURE__ */ new Map();
    return e.forEach((r, s) => {
      t.set(V(s), V(r));
    }), t;
  }
  if (e instanceof Set) {
    const t = /* @__PURE__ */ new Set();
    return e.forEach((r) => {
      t.add(V(r));
    }), t;
  }
  const n = {};
  return Object.keys(e).forEach((t) => {
    n[t] = V(e[t]);
  }), n;
}
function Ee(e, ...n) {
  if (!n.length) return e;
  const t = n.shift();
  return ue(e) && ue(t) && Object.keys(t).forEach((r) => {
    ue(t[r]) ? (e[r] || Object.assign(e, { [r]: {} }), Ee(e[r], t[r])) : Object.assign(e, { [r]: t[r] });
  }), Ee(e, ...n);
}
function ue(e) {
  return !!(e && typeof e == "object" && !Array.isArray(e));
}
function $t(e, n, t = void 0) {
  const r = typeof n == "string" ? n.split(".") : n;
  let s = e;
  for (const o of r) {
    if (s == null)
      return t;
    s = s[o];
  }
  return s === void 0 ? t : s;
}
function Ot(e, n, t) {
  const r = typeof n == "string" ? n.split(".") : n;
  let s = e;
  for (let o = 0; o < r.length - 1; o++) {
    const i = r[o];
    (!(i in s) || s[i] === null || typeof s[i] != "object") && (s[i] = {}), s = s[i];
  }
  return s[r[r.length - 1]] = t, e;
}
function At(e, n) {
  const t = typeof n == "string" ? n.split(".") : n;
  let r = e;
  for (let s = 0; s < t.length - 1; s++) {
    const o = t[s];
    if (!(o in r))
      return e;
    r = r[o];
  }
  return delete r[t[t.length - 1]], e;
}
function Ct(e, n) {
  const t = typeof n == "string" ? n.split(".") : n;
  let r = e;
  for (const s of t) {
    if (!r || !(s in r))
      return false;
    r = r[s];
  }
  return true;
}
function Tt(e, n) {
  return n.reduce((t, r) => (r in e && (t[r] = e[r]), t), {});
}
function Lt(e, n) {
  const t = new Set(n);
  return Object.keys(e).reduce((r, s) => (t.has(s) || (r[s] = e[s]), r), {});
}
function Mt(e = "") {
  const n = Date.now().toString(36), t = Math.random().toString(36).substring(2, 9);
  return `${e}${n}${t}`;
}
function Rt() {
  return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, (e) => {
    const n = Math.random() * 16 | 0;
    return (e === "x" ? n : n & 3 | 8).toString(16);
  });
}
function Je(e) {
  return new Promise((n) => setTimeout(n, e));
}
async function kt(e, n = {}) {
  const {
    attempts: t = 3,
    delay: r = 1e3,
    backoff: s = "exponential",
    onRetry: o = () => {
    }
  } = n;
  let i;
  for (let a = 0; a < t; a++)
    try {
      return await e(a);
    } catch (c) {
      if (i = c, a < t - 1) {
        const u = s === "exponential" ? r * Math.pow(2, a) : r * (a + 1);
        o(c, a + 1, u), await Je(u);
      }
    }
  throw i;
}
function Nt(e, n) {
  const t = /* @__PURE__ */ new Map(), r = (...s) => {
    const o = n ? n(...s) : JSON.stringify(s);
    if (t.has(o))
      return t.get(o);
    const i = e.apply(this, s);
    return t.set(o, i), i;
  };
  return r.cache = t, r.clear = () => t.clear(), r;
}
function It() {
  const e = /* @__PURE__ */ new Map();
  return {
    on(n, t) {
      return e.has(n) || e.set(n, /* @__PURE__ */ new Set()), e.get(n).add(t), () => this.off(n, t);
    },
    off(n, t) {
      e.has(n) && e.get(n).delete(t);
    },
    once(n, t) {
      const r = (...s) => {
        this.off(n, r), t(...s);
      };
      this.on(n, r);
    },
    emit(n, ...t) {
      e.has(n) && e.get(n).forEach((r) => {
        try {
          r(...t);
        } catch (s) {
          console.error(`Error in event handler for "${n}":`, s);
        }
      });
    },
    clear(n) {
      n ? e.delete(n) : e.clear();
    }
  };
}
function xt(e) {
  let n = false;
  return {
    promise: new Promise((r, s) => {
      e.then(
        (o) => n ? s({ cancelled: true }) : r(o),
        (o) => s(n ? { cancelled: true } : o)
      );
    }),
    cancel() {
      n = true;
    }
  };
}
function Pt(e = 1) {
  const n = [];
  let t = 0;
  async function r() {
    if (t >= e || n.length === 0) return;
    t++;
    const { fn: s, resolve: o, reject: i } = n.shift();
    try {
      const a = await s();
      o(a);
    } catch (a) {
      i(a);
    } finally {
      t--, r();
    }
  }
  return {
    add(s) {
      return new Promise((o, i) => {
        n.push({ fn: s, resolve: o, reject: i }), r();
      });
    },
    get pending() {
      return n.length;
    },
    get running() {
      return t;
    },
    clear() {
      n.length = 0;
    }
  };
}
function jt(e) {
  const n = new URLSearchParams(e), t = {};
  return n.forEach((r, s) => {
    s in t ? (Array.isArray(t[s]) || (t[s] = [t[s]]), t[s].push(r)) : t[s] = r;
  }), t;
}
function Dt(e) {
  const n = new URLSearchParams();
  return Object.entries(e).forEach(([t, r]) => {
    r != null && (Array.isArray(r) ? r.forEach((s) => n.append(t, s)) : n.append(t, r));
  }), n.toString();
}
function _t(e) {
  const n = {
    "&": "&amp;",
    "<": "&lt;",
    ">": "&gt;",
    '"': "&quot;",
    "'": "&#39;"
  };
  return e.replace(/[&<>"']/g, (t) => n[t]);
}
function Bt(e, n = 2) {
  if (e === 0) return "0 Bytes";
  const t = 1024, r = ["Bytes", "KB", "MB", "GB", "TB", "PB"], s = Math.floor(Math.log(e) / Math.log(t));
  return parseFloat((e / Math.pow(t, s)).toFixed(n)) + " " + r[s];
}
function Ut(e, n = {}) {
  const { locale: t = "en-US", ...r } = n;
  return new Intl.NumberFormat(t, r).format(e);
}
function qt(e) {
  return e.charAt(0).toUpperCase() + e.slice(1);
}
function zt(e) {
  return e.replace(/[-_\s]+(.)?/g, (n, t) => t ? t.toUpperCase() : "").replace(/^./, (n) => n.toLowerCase());
}
function Ft(e) {
  return e.replace(/([a-z])([A-Z])/g, "$1-$2").replace(/[\s_]+/g, "-").toLowerCase();
}
function Jt(e) {
  return e.replace(/([a-z])([A-Z])/g, "$1_$2").replace(/[\s-]+/g, "_").toLowerCase();
}
const Wt = {
  debounce: Et,
  throttle: St,
  deepClone: V,
  deepMerge: Ee,
  isObject: ue,
  get: $t,
  set: Ot,
  unset: At,
  has: Ct,
  pick: Tt,
  omit: Lt,
  uniqueId: Mt,
  uuid: Rt,
  sleep: Je,
  retry: kt,
  memoize: Nt,
  createEventEmitter: It,
  cancellable: xt,
  createQueue: Pt,
  parseQuery: jt,
  stringifyQuery: Dt,
  escapeHtml: _t,
  formatBytes: Bt,
  formatNumber: Ut,
  capitalize: qt,
  camelCase: zt,
  kebabCase: Ft,
  snakeCase: Jt
};
/**
 * MonkeysJS
 * A lightweight, reactive JavaScript library with built-in HTTP client,
 * declarative DOM binding, caching, WebSockets, and more
 * 
 * @author Yorch / MonkeysCloud
 * @license MIT
 * @version 1.0.0
 */
const Se = {
  // Version
  version: "1.0.0",
  // Core
  ...Y,
  // HTTP
  ...he,
  // WebSocket
  ...ct,
  // DOM
  ...Be,
  // Forms
  ...mt,
  // Storage
  ...vt,
  // Utils
  ...Wt,
  // Install as plugin (for compatibility)
  install(e, n = {}) {
    e.config.globalProperties.$monkeys = Se, e.config.globalProperties.$reactive = Y.reactive, e.config.globalProperties.$ref = Y.ref, e.config.globalProperties.$computed = Y.computed, e.config.globalProperties.$watch = Y.watch, e.config.globalProperties.$http = he.http, n.baseURL && (he.http.defaults.baseURL = n.baseURL);
  }
};
typeof window < "u" && (window.MonkeysJS = Se, window.$m = Se, document.readyState === "loading" && document.addEventListener("DOMContentLoaded", () => {
  Be.autoInit();
}));
